From 3e86c7f124845a9255987a8730d04ba03ae8f7ae Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:40:13 +0800
Subject: [PATCH 21/25] GSI Android 17 compatibility:
 0021-Add-a-prop-to-fix-echo-on-incoming-or-outgoing-calls.patch

Change-Id: I6cf8e90a023db53017196669a6d158b882953712
---
 .../common/managerdefinitions/src/Serializer.cpp         | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp b/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
index 26edcaa..059e4ad 100644
--- a/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
+++ b/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
@@ -669,6 +669,9 @@ std::variant<status_t, RouteTraits::Element> PolicySerializer::deserialize<Route
     // This fixes broken mic while video record on some Exynos devices
     bool disableBackMic = property_get_bool("persist.sys.phh.disable_back_mic", false);
 
+    // This may fix echo or non-working audio on calls
+    bool disableVoiceCallIn = property_get_bool("persist.sys.phh.disable_voice_call_in", false);
+
     std::string sourcesAttr = getXmlAttribute(cur, Attributes::sources);
     if (sourcesAttr.empty()) {
         ALOGE("%s: No %s found", __func__, Attributes::sources);
@@ -685,10 +688,16 @@ std::variant<status_t, RouteTraits::Element> PolicySerializer::deserialize<Route
             if (source == NULL) {
                 source = ctx->findPortByTagName(trim(devTag));
             }
+
             if (disableBackMic && strcmp(devTag, "Built-In Back Mic") == 0) {
                 ALOGW("Skipping route source \"%s\" as it breaks video recording mic", devTag);
                 source = NULL;
+            } else if (disableVoiceCallIn && strcmp(devTag, "Voice Call In") == 0 &&
+                      (sinkAttr == "voice tx" || sinkAttr == "voice_tx")) {
+                ALOGW("Skipping route source \"%s\" as it breaks audio on calls", devTag);
+                source = NULL;
             }
+
             if (source == NULL) {
                 if (false && !mIgnoreVendorExtensions) {
                     ALOGE("%s: no source found with name \"%s\"", __func__, devTag);
